text/0000-zero-sized-references.md

```diff
@@ -0,0 +1,142 @@
+A struct with a reference to itself will not know it's size until after it knows the size to the reference.
+
+However, a couple of notes:
+
+```rust
+struct A<'a> (&'a A<'a>);
+```
+This struct cannot be instantiated, because the very first instance of it requires an instance of it to already exist.
+
+```rust
+enum A<'a> {
+    ZeroSized,
+    SelfRef (&'a A<'a>),
+}
+```
+The moment an enum has more than a single value, it cannot be Zero-Sized. Otherwise it isn't different than a struct.
+
+Therefore, I propose to assume that whenever you find a self reference (or multiple types referencing in a loop),
+decide the reference is not Zero-Sized, since there most likely WILL be other data somewhere in the chain.
+This is most relevant to unions, which could have self-references and be instantiated at the same time.
+
+### `*const` and `*mut` pointers
+[pointers]: #pointers
+
+It is possible to convert a reference to a pointer. Currently, a reference to ZST points to an arbitrary location,
+and when converting to a pointer the pointer recieves that arbitrary location.
+
+After this change, the reference will not hold any data. I propose that whenever a ZST reference is converted to a pointer,
+a warning/error be issued ("Warning: taking the address of a Zero-Sized Type is meaningless") and the pointer will recieve an address
+with the same algorithm that assigned an address to the reference in our current implementation.
+
+The purpose is to not break current code that might do this. We probably don't want to assign Null since for pointers it has
+a meaning that the value doesn't exist, which is different than "exist but no data".
+
+Converting in the other direction, the value of the pointer will be silently dropped - that value never had a meaning in the first place.
+
+# How We Teach This
+[how-we-teach-this]: #how-we-teach-this
+
+For most rust users, this change will be invisible. Thier code will just become a tiny bit smaller.
+
+Users of unsafe rust might encounter this case. Therefore there should probably be a note in the nomicon that references might be optimized away for ZST.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+### Breaking code
+
+Any place that assumed a reference holds a pointer might introduce bugs.
+
+FFI might behave differently than now, breaking code (see Unresolved questions).
+
+# Alternatives
+[alternatives]: #alternatives
+
+The system that is now works well, and does not have to be changed.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+### FFI + Escape mechanism
+
+Do we need a way to remove the optimization? For example, if we have
+```rust
+struct ExternalStruct;
+
+let x: &mut ExternalStruct = ffi_function();
+other_ffi_function(x);
+```
+We might want to represent a pointer we got from FFI and checked it's not null as a reference to an object.
+Since that object is not accessable directly we represent it as an empty struct.
```

Comment:
afaik that's currently the best way to represent external structures, see the discussion on the `extern type` RFC